File: scripts/generate-icons.js
/**
 * generate-icons.js — minimal PNG icon generator (zero dependencies)
 * Generates valid PNG files with a styled "S" letter icon.
 */
const fs   = require('fs')
const path = require('path')
const zlib = require('zlib')

const OUT_DIR = path.join(__dirname, '../public/icons')
fs.mkdirSync(OUT_DIR, { recursive: true })

// ── Minimal PNG encoder ───────────────────────────────────────────────────────
function uint32BE(n) {
  const b = Buffer.allocUnsafe(4)
  b.writeUInt32BE(n, 0)
  return b
}

function chunk(type, data) {
  const typeB = Buffer.from(type, 'ascii')
  const len   = uint32BE(data.length)
  const crcBuf = Buffer.concat([typeB, data])
  let crc = 0xffffffff
  for (const byte of crcBuf) {
    crc ^= byte
    for (let i = 0; i < 8; i++) crc = (crc & 1) ? (crc >>> 1) ^ 0xedb88320 : crc >>> 1
  }
  crc ^= 0xffffffff
  const crcB = uint32BE(crc >>> 0)
  return Buffer.concat([len, typeB, data, crcB])
}

function makePNG(pixels, size) {
  // pixels: Uint8Array of RGBA values, row by row
  const IHDR = chunk('IHDR', Buffer.concat([
    uint32BE(size), uint32BE(size),
    Buffer.from([8, 2, 0, 0, 0]),  // bit depth 8, color type RGB (2)
  ]))
  // Actually use RGBA (color type 6)
  const ihdrData = Buffer.concat([
    uint32BE(size), uint32BE(size),
    Buffer.from([8, 6, 0, 0, 0]),
  ])
  const IHDRchunk = chunk('IHDR', ihdrData)

  // Raw image data: filter byte (0) + row RGBA data
  const rowSize   = size * 4
  const rawRows   = Buffer.allocUnsafe(size * (rowSize + 1))
  for (let y = 0; y < size; y++) {
    rawRows[y * (rowSize + 1)] = 0  // filter type None
    for (let x = 0; x < size; x++) {
      const srcOff = (y * size + x) * 4
      const dstOff = y * (rowSize + 1) + 1 + x * 4
      rawRows[dstOff]     = pixels[srcOff]
      rawRows[dstOff + 1] = pixels[srcOff + 1]
      rawRows[dstOff + 2] = pixels[srcOff + 2]
      rawRows[dstOff + 3] = pixels[srcOff + 3]
    }
  }
  const compressed = zlib.deflateSync(rawRows)
  const IDATchunk  = chunk('IDAT', compressed)
  const IENDchunk  = chunk('IEND', Buffer.alloc(0))
  const sig        = Buffer.from([137,80,78,71,13,10,26,10])
  return Buffer.concat([sig, IHDRchunk, IDATchunk, IENDchunk])
}

// ── Rasterize the "S" icon ────────────────────────────────────────────────────
function renderIcon(size) {
  const pixels = new Uint8Array(size * size * 4)

  const cx = size / 2
  const cy = size / 2
  const r  = size * 0.42   // outer radius of rounded rect

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const off = (y * size + x) * 4
      const dx  = x - cx
      const dy  = y - cy

      // Background: superellipse (rounded square)
      const n   = 4
      const val = Math.pow(Math.abs(dx / r), n) + Math.pow(Math.abs(dy / r), n)
      if (val <= 1) {
        // Gradient: indigo(top-left) → purple(bottom-right)
        const t = (x / size + y / size) / 2
        pixels[off]     = Math.round(0x4f + (0x7c - 0x4f) * t)  // R
        pixels[off + 1] = Math.round(0x46 + (0x3a - 0x46) * t)  // G
        pixels[off + 2] = Math.round(0xe5 + (0xed - 0xe5) * t)  // B
        pixels[off + 3] = 255

        // "S" letter: simple mask-based approach
        const nx = dx / r + 0.5
        const ny = dy / r + 0.5

        // Top bar of S
        if (ny > 0.05 && ny < 0.18 && nx > -0.25 && nx < 0.35) {
          pixels[off] = 255; pixels[off+1] = 255; pixels[off+2] = 255
        }
        // Middle bar of S
        if (ny > 0.41 && ny < 0.54 && nx > -0.35 && nx < 0.25) {
          pixels[off] = 255; pixels[off+1] = 255; pixels[off+2] = 255
        }
        // Bottom bar of S
        if (ny > 0.77 && ny < 0.90 && nx > -0.35 && nx < 0.25) {
          pixels[off] = 255; pixels[off+1] = 255; pixels[off+2] = 255
        }
        // Left side top
        if (nx > -0.45 && nx < -0.30 && ny > 0.05 && ny < 0.54) {
          pixels[off] = 255; pixels[off+1] = 255; pixels[off+2] = 255
        }
        // Right side bottom
        if (nx > 0.20 && nx < 0.35 && ny > 0.41 && ny < 0.90) {
          pixels[off] = 255; pixels[off+1] = 255; pixels[off+2] = 255
        }
      }
    }
  }
  return pixels
}

for (const size of [16, 48, 128]) {
  const pixels = renderIcon(size)
  const png    = makePNG(pixels, size)
  const file   = path.join(OUT_DIR, `icon${size}.png`)
  fs.writeFileSync(file, png)
  console.log(`✓  icon${size}.png  (${png.length} bytes)`)
}

File: scripts/generate-icons.py
"""
Generate minimal PNG icons for StyleSnap extension.
Pure stdlib — no Pillow needed.
"""
import struct, zlib, math, os

OUT_DIR = os.path.join(os.path.dirname(__file__), '..', 'src', 'assets', 'icons')
os.makedirs(OUT_DIR, exist_ok=True)

def make_png(pixels: bytes, size: int) -> bytes:
    """Encode raw RGBA pixels as PNG."""
    def crc32(data):
        return zlib.crc32(data) & 0xffffffff

    def chunk(tag: bytes, data: bytes) -> bytes:
        c = struct.pack('>I', len(data)) + tag + data
        return c + struct.pack('>I', crc32(tag + data))

    ihdr = struct.pack('>IIBBBBB', size, size, 8, 6, 0, 0, 0)
    # Build raw scanlines: filter byte 0 + RGBA row
    raw = bytearray()
    for y in range(size):
        raw.append(0)
        raw.extend(pixels[y * size * 4 : (y + 1) * size * 4])
    idat_data = zlib.compress(bytes(raw), 9)

    return (
        b'\x89PNG\r\n\x1a\n'
        + chunk(b'IHDR', ihdr)
        + chunk(b'IDAT', idat_data)
        + chunk(b'IEND', b'')
    )

def render_icon(size: int) -> bytes:
    """Rasterize a rounded-square icon with a white 'S' letter."""
    pixels = bytearray(size * size * 4)
    cx = cy = size / 2

    for y in range(size):
        for x in range(size):
            off = (y * size + x) * 4
            dx = x - cx + 0.5
            dy = y - cy + 0.5

            # Superellipse background (n=5 for smooth rounded square)
            r = size * 0.40
            n = 5
            val = (abs(dx / r) ** n) + (abs(dy / r) ** n)
            if val > 1.0:
                continue  # transparent

            # Indigo→purple gradient
            t = (x / size * 0.5 + y / size * 0.5)
            R = int(0x4f + (0x7c - 0x4f) * t)
            G = int(0x46 + (0x3a - 0x46) * t)
            B = int(0xe5 + (0xed - 0xe5) * t)
            pixels[off]     = R
            pixels[off + 1] = G
            pixels[off + 2] = B
            pixels[off + 3] = 255

            # Normalised coords for letter drawing (0..1 within bounding circle)
            nx = dx / r  # -1..1
            ny = dy / r

            # Draw a simple "S" shape using rectangles in normalised space
            # All coords: nx in (-1, 1), ny in (-1, 1), top=-1, bottom=+1
            def rect(x0, y0, x1, y1):
                return x0 <= nx <= x1 and y0 <= ny <= y1

            # Top horizontal bar
            if rect(-0.42, -0.78, 0.42, -0.48):
                pixels[off:off+4] = [255,255,255,255]; continue
            # Upper-left vertical
            if rect(-0.56, -0.78, -0.24, -0.05):
                pixels[off:off+4] = [255,255,255,255]; continue
            # Middle bar
            if rect(-0.42, -0.14, 0.42,  0.16):
                pixels[off:off+4] = [255,255,255,255]; continue
            # Lower-right vertical
            if rect(0.24,  0.05, 0.56,  0.78):
                pixels[off:off+4] = [255,255,255,255]; continue
            # Bottom bar
            if rect(-0.42,  0.48, 0.42,  0.78):
                pixels[off:off+4] = [255,255,255,255]; continue

    return make_png(bytes(pixels), size)

for size in (16, 48, 128):
    data = render_icon(size)
    path = os.path.join(OUT_DIR, f'icon{size}.png')
    with open(path, 'wb') as f:
        f.write(data)
    print(f'  icon{size}.png  {len(data):,} bytes')

print('Done.')
